Preview: Side panel (#2410)
strapi/documentation: docusaurus/docs/dev-docs/preview.md

}
```

### 5. [Front end] Allow the front-end to be embedded {#5-allow-frontend-embed}

On the Strapi side, [the `allowedOrigins` configuration parameter](#allowed-origins) allows the admin panel to load the front-end window in an iframe. But allowing the embedding works both ways, so on the front-end side, you also need to allow the window to be embedded in Strapi's admin panel.

This requires the front-end application to have its own header directive, the CSP `frame-ancestors` directive. Setting this directive up depends on how your website is built. For instance, setting this up in Next.js requires a middleware configuration (see [Next.js docs](https://nextjs.org/docs/app/building-your-application/configuring/content-security-policy)).

### 6. [Front end] Detect changes in Strapi and refresh the front-end {#6-refresh-frontend}

Strapi emits a `strapiUpdate` message to inform the front end that data has changed. 

> To track this, within your front-end application, add an event listener to listen to events posted through [the `postMessage()` API](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage). The listener needs to filter through messages and react only to Strapi-initiated messages, then refresh the iframe.

Reviewer: >add an event listener to listen to events posted through [the `postMessage()` API]

Maybe it's overly specific but we could specify "posted _on the window object_"

Reviewer: >then refresh the iframe

This seems to imply a hard refresh, maybe say "then refresh the iframe content"?

> To track this, within your front-end application, add an event listener to listen to events posted through [the `postMessage()` API](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage). The listener needs to filter through messages and react only to Strapi-initiated messages, then refresh the iframe.

Author: ```suggestion
To track this, within your front-end application, add an event listener to listen to events posted through [the `postMessage()` API](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) on the `window` object. The listener needs to filter through messages and react only to Strapi-initiated messages, then refresh the iframe content.
```

        router.refresh();
      }
    };
    
    // Add the event listener
    window.addEventListener("message", handleMessage);
    
    // Cleanup the event listener on unmount
    return () => {
      window.removeEventListener("message", handleMessage);
    };
  }, [router]);
  
  // ...
}
```

</TabItem>
<TabItem value="ts" label="TypeScript" >

```tsx title="next/app/path/to/your/front/end/logic.tsx" {6-17}
export default function MyClientComponent({
  //…
  const router = useRouter();

  useEffect(() => {
    const handleMessage = async (message: MessageEvent<any>) => {
      if (
        // Filters events emitted through the postMessage() API
        message.origin === process.env.NEXT_PUBLIC_API_URL &&
        message.data.type === "strapiUpdate"
      ) { // Recommended way to refresh with Next.js
        router.refresh();
      }
    };

    // Add the event listener
    window.addEventListener("message", handleMessage);

    // Cleanup the event listener on unmount
    return () => {
      window.removeEventListener("message", handleMessage);
    };
  }, [router]);

  // …
})
```

</TabItem>

</Tabs>

<details>
<summary>Caching in Next.js:</summary>

> The persistence of [cache in the back end of Next.js](https://nextjs.org/docs/app/building-your-application/caching) may need additional steps where you would need to invalidate cache by sending an API call from the client side of Next to the server side which will handle the revalidation logic. Please refer to Next.js documentation for details, for instance with the [revalidatePath() method](https://nextjs.org/docs/app/building-your-application/caching#revalidatepath).

Reviewer: The first sentence here could use more punctuation, it's a bit long to read. Also I wouldn't say "in the backend" in the beginning, especially since you're linking to Next's overall caching strategy

> The persistence of [cache in the back end of Next.js](https://nextjs.org/docs/app/building-your-application/caching) may need additional steps where you would need to invalidate cache by sending an API call from the client side of Next to the server side which will handle the revalidation logic. Please refer to Next.js documentation for details, for instance with the [revalidatePath() method](https://nextjs.org/docs/app/building-your-application/caching#revalidatepath).

Author: ```suggestion
In Next.js, [cache persistence](https://nextjs.org/docs/app/building-your-application/caching) may require additional steps. You might need to invalidate the cache by making an API call from the client side to the server, where the revalidation logic will be handled. Please refer to Next.js documentation for details, for instance with the [revalidatePath() method](https://nextjs.org/docs/app/building-your-application/caching#revalidatepath).
```
